pydantic/_internal/_discriminated_union.py

```diff
@@ -217,7 +188,7 @@ def _apply_to_root(self, schema: core_schema.CoreSchema) -> core_schema.CoreSche
             # was flattened by pydantic_core.
             # However, it still may make sense to apply the discriminator to this schema,
             # as a way to get discriminated-union-style error messages, so we allow this here.
-            schema = core_schema.union_schema([schema])
+            schema = core_schema.union_schema([schema.copy()])
```

Comment:
Cleaned it up. Removed this change, instead for simplicity do the shallow copy in caller side in `finalize_schema_defs_inplace`. It might do some unnecessary copying in above cases `if schema['type'] == 'nullable'` or `if schema['type'] == 'definitions'` but they feel quite rare.